Individual data download policies via dataset-specific configuration YAML

claire-simpson opened the pull request:
This PR introduces per-dataset download policies that (1) let us block individual files (by name/regex/extension) and (2) allow downloading subfolders (as inferred `.zip`s). That is, we can now create dataset-specific config YAMLs that can allow subfolders to be downloaded (via zip) and individual files to be blocked for download. Note that in order for subfolders to be downloadable, a zipfile (with the same name as the subfolder) must exist in GCS adjacent to the subfolder itself.

Another option I explored was auto-detecting any zipfiles that have the same name and location as a subfolder, and if such a file was found, automatically adding it as a download option. I ended up not going for this method because it was slower to check that the file exists on GCS and I wasn't sure that most datasets would want this option by default, as likely most subfolders won't be downloable/have an associated zip. Curious if others have different opinions on this! 

I currently have it set up for the config YAMLs to live at `data.naturalcapitalproject.stanford.edu/src/ckanext-natcap/ckanext/natcap/dataset_configs/<dataset-title-or-name>.yml`, but I'm open to other suggestions as well.

In addition, I'd appreciate feedback on if the config yaml example I have for 3Ps is intuitive. If/when we have to make more of these dataset config files, is there an easier/better way that this yml could be designed?

Fixes #165

phargogh commented on src/ckanext-natcap/ckanext/natcap/plugin.py:
```diff
@@ -240,6 +243,121 @@ def natcap_update_mappreview(context, package):
     #NatcapPlugin._after_dataset_update(context, package_data)
 
 
+def _load_download_rules_for(pkg):
```
The truthy comparator a little later in the function body is pretty interesting ... could you describe what we should expect `pkg` to be?  Is any falsey value allowed (e.g. `{}` vs `None`) or are we expecting a specific falsey variable?

claire-simpson replied:
I've added docstrings for every new function so hopefully that should clarify many of these questions. `pkg` should be a CKAN package metadata dict, and I don't know why it would be falsey as we're getting this from `c.pkg_dict` so I can remove that bit (just being cautious!)

phargogh commented on src/ckanext-natcap/ckanext/natcap/plugin.py:
```diff
@@ -240,6 +243,121 @@ def natcap_update_mappreview(context, package):
+
+    Returns:
+        dict: {'allowed': bool, 'reason': '...'}
+    """
```
I think I'm inferring from the function itself that `source` is either a dict or an object(?) or maybe something else.  It would be really helpful to define and document the contract up front if we can and then stick to it!

phargogh commented on src/ckanext-natcap/ckanext/natcap/plugin.py:
```diff
@@ -240,6 +243,121 @@ def natcap_update_mappreview(context, package):
+def _match_rule(m, relpath, base, ext):
+    pg = m.get('path_glob')
+    if pg and not fnmatch.fnmatch(relpath, pg):
+        return False
```
I see the stdlib `fnmatch` (which I didn't know about! always a good day to learn something new about the stdlib), but this also kind of only appears to match filenames, not necessarily directories, or nested directories.  My understanding of the stdlib `glob` module is that this would be able to support nested directories.  Maybe that would provide a little more functionality for not so much of a change to this function?

claire-simpson replied:
`fnmatch` can also do pattern matching on directories! I've updated `_match_rule` to enable this functionality and added a rule to the 3Ps config yaml to demonstrate. 

I neglected to apply on this functionality before because the `name` of the `source` is just the basename, rather than the full filepath, so I didn't think I could match directories (since directory structure info wasn't being passed to the function). I now have `_match_rule` checking the file download `url` though, which seems to work fine!

empavia commented:
There are 2 scenarios this should be relevant for:
1. We have a list of files we only want to be downloaded at the folder level, not individually. This is exemplified by the Colombia Data Package as the impetus for this PR. 
2. We have files that we do not want to be downloadable, period. This is exemplified by the [WorldClim precip](https://data-staging.naturalcapitalproject.org/dataset/sts-fe7d4eeb78b5a6e7b99f166c1e28965114ff3f6277c53618df5447421fb012fe) data on staging. We want to visualize the layers, but the only link that should be clickable is the link to the external download page ("html worldclim 21").

As far as issue 1 goes, the proposed solution looks good to me. I am not 100% sure whether auto-detecting a zipfile is the right way to go. On one hand, it sounds like it may not be practical due to the time it takes. However, it may make sense if we decide that only subfolders that should be downloaded as a whole have a zipfile. The issue I have with this is not being sure whether we can assume a zipfile's presence should result in this. All to say, we may want to manually add this to the config for now, but keep an idea like this on the back burner. As far as the config file itself goes, it makes logistical sense as far as I can tell. I am not familiar with the implementation of a file such as this, but it feels intuitive, especially with consideration of the comments made on the file itself. 

For issue 2, I believe the config file will work here, as well. We could ideally change all download rules to `allow: false` as I understand it. An additional step for this may be adding GCP authentication, as it would be best to keep the files private on GCP. Considering the authentication tokens would be the same for all private datasets, we may just have that configured elsewhere. 

Hopefully this all makes sense! Let me know if you need more context for what we need working on the UI side here.

claire-simpson commented on src/ckanext-natcap/ckanext/natcap/plugin.py:
```diff
@@ -240,6 +243,121 @@ def natcap_update_mappreview(context, package):
+    base = name.rstrip('/').rsplit('/', 1)[-1]
+    ext = ''
+    if '.' in base and not base.endswith('.'):
+        ext = '.' + base.rsplit('.', 1)[-1]
```
Oops yes good call

phargogh commented on src/ckanext-natcap/ckanext/natcap/plugin.py:
```diff
@@ -244,51 +244,87 @@ def natcap_update_mappreview(context, package):
+    try:
+        if path.exists(potential_config_path):
+            with open(potential_config_path, 'r') as f:
+                data = yaml.safe_load(f) or {}
```
I don't think I commented on this before, but excellent idea to use `safe_load` here, even though we _should_ (hopefully!) be operating on trusted data.

claire-simpson commented on src/ckanext-natcap/ckanext/natcap/dataset_configs/People, Planet, Prosperity Pilot Country Data Package - Colombia.yml:
```diff
@@ -0,0 +1,31 @@
+  - match:
+      name_regex: "(?i).*p_EPSG32618_monthly.*"
+    allow: false
+    reason: "Monthly precip files are not downloadable"
```
Ok I don't actually think that `ckan` supports python 3.13 (https://docs.ckan.org/en/2.11/changelog.html#v-2-11-4-2025-10-29). It looks like as of ckan 2.11, up to python 3.12 support was added. Unless you have an issue with `fnmatch`, I think I'll proceed with that!

claire-simpson commented:
Status update: I think I've addressed all requested changes! I added an example template config as well as a WorldClim config. The latter does not have any effect yet, but I believe this has more to do with how WorldClim dataset's metadata is configured or the datahub's ability to handle external datasets than the config yaml, as the `get_file_downloadability` function is not being called at all for this dataset (additionally, the filenames for the geotiffs in this dataset are off due to the `.` in the middle of each filename). I will try to look into this a bit more, but my opinion is this issue is likely a bit beyond the scope of this PR and I don't think it should be a limiting factor here.

The last outstanding issue is the alignment of filenames and download button. I have not been able to replicate the css/visualization issue that Megan described; the alignment of files and the download button appear correctly justified for me.
